warehouse/manage/views.py

```diff
@@ -288,6 +300,121 @@ def delete_account(self):
         return logout(self.request)
 
 
+@view_defaults(
+    route_name="manage.account.totp-provision",
+    renderer="manage/account/totp-provision.html",
+    uses_session=True,
+    require_csrf=True,
+    require_methods=False,
+    permission="manage:user",
+)
+class ProvisionTOTPViews:
+    def __init__(self, request):
+        self.request = request
+        self.user_service = request.find_service(IUserService, context=None)
+        self.token_service = request.find_service(ITokenService, name="totp_provision")
+
+    @property
+    def default_response(self):
+        return HTTPSeeOther(self.request.route_path("manage.account"))
+
+    @view_config(request_method="GET")
+    def totp_provision(self):
+        two_factor = self.user_service.get_two_factor(self.request.user.id)
+        if two_factor.totp_secret:
+            self.request.session.flash("TOTP already provisioned.", queue="error")
+            return self.default_response
+
+        try:
+            token_data = self.token_service.loads(
+                self.request.cookies.get(TWO_FACTOR_TOTP_PROVISION_COOKIE)
+            )
+        except TokenException:
+            token_data = {
+                "totp_secret": urlsafe_b64encode(otp.generate_totp_secret()).decode()
+            }
+            self.request.response.set_cookie(
```

Comment:
Isn't this putting the TOTP secret in the cookie store of the browser in cleartext? What's the rationale for doing that versus storing it in the server side session data? If it needs to be in a cookie should we be encrypting it at least?

I know there was some discussion related to this so I am probably just missing context here :)

Reply:
Yup! The context here is the dual-state discussion + my commit message in https://github.com/pypa/warehouse/pull/5567/commits/37ed90dce638d180079bcf653a7b6e0acbe1ce09.

In sum: I used a cookie here because I didn't like my workaround of having two separate TOTP columns + a potentially stale TOTP secret, and because I wasn't familiar with the Warehouse helpers/idioms for storing session data on the server-side. If you have any pointers on that, I'd greatly appreciate them!

As for encrypting: TOTP is fully symmetric, so we need to expose the TOTP secret entirely anyways: it's in the provisioning URI as a base32-encoded string. The cookie store is definitely a wider security scope though (especially if the cookie somehow outlives the provisioning flow), so this is another good reason to switch to server-side state.